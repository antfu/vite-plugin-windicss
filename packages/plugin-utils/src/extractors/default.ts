import type { ExtractorResultDetailed } from 'windicss/types/interfaces'
import { regexClassSplitter, validClassName, regexHtmlTag, regexAttributifyItem } from '../regexes'

export function DefaultExtractor(code: string, id?: string): ExtractorResultDetailed {
  if (id?.endsWith('.css') || id?.endsWith('.postcss')) {
    return {
      classes: [],
      tags: [],
    }
  }

  const tags = Array.from(code.matchAll(regexHtmlTag))
  const tagNames = tags.map(i => i[1])

  return {
    tags: tagNames,
    get classes() {
      return code
        .split(regexClassSplitter)
        .filter(validClassName)
    },
    get attributes() {
      const attrRanges: [number, number][] = []
      const attributes: ExtractorResultDetailed['attributes'] = {
        names: [],
        values: [],
      }

      const attributesBlocklist = ['class', 'className']
      const tagsBlocklist = ['meta', 'script', 'style', 'link']

      tags
        .filter(i => !tagsBlocklist.includes(i[1]))
        .forEach((i) => {
          return Array.from(i[2].matchAll(regexAttributifyItem) || [])
            .forEach((match) => {
              let name = match[1]
              const [full,,, value] = match
              // remove vue binding
              name = name.replace(/^(:|v-bind:)/, '')
              if (attributesBlocklist.includes(name))
                return
              attributes.names.push(name)
              attributes.values.push(value)
              if (match.index != null)
                attrRanges.push([match.index, match.index + full.length])
            })
        })

      // Disable this feature for now as we need to find a way to avoid false-negative
      //
      // attributes.classes = Array.from(code.matchAll(regexQuotedString))
      //   .flatMap((m) => {
      //     if (m.index != null && attrRanges.some(([start, end]) => m.index! >= start && m.index! <= end))
      //       return []
      //     return (m[2] || '').split(regexClassSplitter)
      //   })
      //   .filter(validClassName)

      return attributes
    },
  }
}
