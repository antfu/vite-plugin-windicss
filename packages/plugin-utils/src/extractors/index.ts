export * from './default'
export * from './helper'
export * from './pug'
