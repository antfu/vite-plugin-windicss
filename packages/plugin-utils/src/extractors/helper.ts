import { extname } from 'path'
import { Extractor } from 'windicss/types/interfaces'
import { DefaultExtractor } from './default'
import { PugExtractor } from './pug'

export function getDefaultExtractors() {
  const extractors: Extractor[] = []

  // auto detect pug
  try {
    require.resolve('pug')
    extractors.push({
      extractor: PugExtractor,
      extensions: ['vue', 'pug'],
    })
  }
  catch (e) {}

  return extractors
}

export async function applyExtractors(code: string, id?: string, extractors: Extractor[] = [], defaultExtract = DefaultExtractor) {
  let ext = id ? extname(id) : '*'
  if (ext[0] === '.')
    ext = ext.slice(1)
  for (const { extractor, extensions } of extractors) {
    if (extensions.includes(ext))
      return extractor(code, id)
  }
  return defaultExtract(code, id)
}
