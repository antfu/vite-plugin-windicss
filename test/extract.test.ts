import { DefaultExtractor } from '../packages/plugin-utils/src/extractors/default'

describe('extract', () => {
  it('attributify', () => {
    expect(DefaultExtractor(`
      <div
        p="x-4 y-2 md:x-2 lg:hover:x-8"
        border="~"
        sm="bg-red-500 bg-opacity-50 hover:bg-green-300"
        sm:hover="text-red-500 text-lg focus:bg-white"
dark:border="~ red-400"
        @click="a()"
        class="text-red-400"
      />
    `),
    ).toMatchSnapshot()
  })

  // #162, #193
  it('partial 1', () => {
    expect(DefaultExtractor(`
      /* I'll fix this */
      const class = 'p-2'
    `).classes,
    ).toContain('p-2')
  })

  it('partial 2', () => {
    expect(DefaultExtractor(`
      <!-- I'll fix this -->
      <p class='p-2'></p>
    `).classes,
    ).toContain('p-2')
  })

  it('partial 3', () => {
    expect(DefaultExtractor(`
      <!-- I'll fix this -->
      <p class="p-2"></p>
      <!-- I'll fix that -->
    `).classes,
    ).toContain('p-2')
  })

  it('partial 4', () => {
    expect(DefaultExtractor(`
      <p class="<sm:text-primary"></p>
    `).classes,
    ).toContain('<sm:text-primary')
  })
})
