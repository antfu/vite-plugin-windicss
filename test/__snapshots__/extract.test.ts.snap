// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract attributify 1`] = `
Object {
  "attributes": Object {
    "names": Array [
      "p",
      "border",
      "sm",
      "sm:hover",
      "dark:border",
    ],
    "values": Array [
      "x-4 y-2 md:x-2 lg:hover:x-8",
      "~",
      "bg-red-500 bg-opacity-50 hover:bg-green-300",
      "text-red-500 text-lg focus:bg-white",
      "~ red-400",
    ],
  },
  "classes": Array [
    "<div",
    "x-4",
    "y-2",
    "md:x-2",
    "lg:hover:x-8",
    "bg-red-500",
    "bg-opacity-50",
    "hover:bg-green-300",
    "text-red-500",
    "text-lg",
    "focus:bg-white",
    "red-400",
    "a()",
    "text-red-400",
  ],
  "tags": Array [
    "div",
  ],
}
`;
